src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs

```diff
@@ -170,121 +111,72 @@ private void AddAccessControlHeaders(HttpRequestMessage request, HttpResponseMes
 				if (Request.Method == HttpMethod.Post)
 				{
 					Dictionary<Guid, List<int>> preMap = new Dictionary<Guid, List<int>>();
+					List<DmlResult> finalResult = new List<DmlResult>();
 
-				List<DmlResult> finalResult = new List<DmlResult>();
-
-				for (int i = 0; i < obj.Length; i++)
-				{
-					List<int> templist = new List<int>();
-					if (obj[i].PartitionId == Guid.Empty)
-					{
-						obj[i].PartitionId = await GetRandomPartitionId(serviceUri);
-					}
-					if (preMap.ContainsKey(obj[i].PartitionId))
-					{
-						templist = preMap[obj[i].PartitionId];
-						templist.Add(i);
-						preMap[obj[i].PartitionId] = templist;
-					}
-					else
-					{
-						templist.Add(i);
-						preMap[obj[i].PartitionId] = templist;
-					}
-				}
-
-				foreach (Guid mypid in preMap.Keys)
-				{
-					//Fetch partition proxy.
-					var proxy = await GetServiceProxyForPidAsync<IQueryableService>(serviceUri, mypid).ConfigureAwait(false);
-
-					foreach (int myref in preMap[mypid])
-					{
-						string keyquoted = JsonConvert.SerializeObject(obj[myref].Key,
-							new JsonSerializerSettings { StringEscapeHandling = StringEscapeHandling.EscapeNonAscii });
-
-						string valuequoted = JsonConvert.SerializeObject(obj[myref].Value,
-							new JsonSerializerSettings { StringEscapeHandling = StringEscapeHandling.EscapeNonAscii });
-
-						DmlResult tempResult = new DmlResult();
-						tempResult.Key = obj[myref].Key;
-						tempResult.PartitionId = mypid;
-						tempResult.Status = await proxy.AddAsync(collection, keyquoted, valuequoted);
-						finalResult.Add(tempResult);
-					}
-				}
-				content = JsonConvert.SerializeObject(finalResult);
-			}
-				// Return response, with appropriate CORS headers.
-			var response = new HttpResponseMessage { Content = new StringContent(content, Encoding.UTF8, "application/json") };
-			AddAccessControlHeaders(Request, response);
-			return new ResponseMessageResult(response);
-		
-		
-
-			//	return Ok(finalResult);
-			}
-			catch (Exception e)
-			{
-				return HandleException(e, serviceUri);
-			}
-		}
-
-		protected async Task<IHttpActionResult> UpdateAsync(string application, string service, string collection,
-			ValueViewModel[] obj)
-		{
-			var serviceUri = GetServiceUri(application, service);
-			try
-			{
-				string content = string.Empty;
-				if (Request.Method == HttpMethod.Put)
-				{
-					Dictionary<Guid, List<int>> preMap = new Dictionary<Guid, List<int>>();
-				List<DmlResult> finalResult = new List<DmlResult>();
-
-				for (int i = 0; i < obj.Length; i++)
-				{
-					List<int> templist = new List<int>();
-					if (preMap.ContainsKey(obj[i].PartitionId))
-					{
-						templist = preMap[obj[i].PartitionId];
-						templist.Add(i);
-						preMap[obj[i].PartitionId] = templist;
-					}
-					else
+					for (int i = 0; i < obj.Length; i++)
```

Comment:
Sometimes partitionID does not exist in incoming data, in that case I generate a random partitionId and assign it right there, and then grouping happens. And for that we anyways have to loop at least  once through the data, so we might as well map it in the same loop.